elements/lisk-api-client/src/ws_client.ts

```diff
@@ -13,15 +13,183 @@
  *
  */
 
-import { APIClient } from './api_client';
-import { Channel } from './types';
+import * as WebSocket from 'isomorphic-ws';
+import { EventEmitter } from 'events';
 
-export const createWSClient = async (url: string): Promise<APIClient> => {
-	// FIXME: requires real implementation
-	const channel = ({ url } as unknown) as Channel;
-	await channel.connect();
-	const client = new APIClient(channel);
-	await client.init();
+const CONNECTION_TIMEOUT = 2000;
+const ACKNOWLEDGMENT_TIMEOUT = 2000;
+const RESPONSE_TIMEOUT = 3000;
 
-	return client;
+const timeout = async <T = void>(ms: number, message?: string): Promise<T> =>
+	new Promise((_, reject) => {
+		const id = setTimeout(() => {
+			clearTimeout(id);
+			reject(new Error(message ?? `Timed out in ${ms}ms.`));
+		}, ms);
+	});
+
+type EventCallback = (...args: any[]) => void;
+
+interface WSResponse {
+	id?: number | string | null;
+	method: string;
+	params?: object;
+	result?: any;
+	error?: any;
+}
+
+interface Defer<T> {
+	promise: Promise<T>;
+	resolve: (result: T) => void;
+	reject: (error?: Error) => void;
+}
+
+const defer = <T>(): Defer<T> => {
+	let resolve!: (res: T) => void;
+	let reject!: (error?: Error) => void;
+
+	const promise = new Promise<T>((_resolve, _reject) => {
+		resolve = _resolve;
+		reject = _reject;
+	});
+
+	return { promise, resolve, reject };
 };
+
+export class WSClient {
+	public isAlive = false;
+	private readonly _url: string;
+	private _ws?: WebSocket;
+	private _requestCounter = 0;
+	private _pendingRequests: {
+		[key: number]: Defer<any>;
+	} = {};
+	private readonly _emitter: EventEmitter;
+
+	public constructor(url: string) {
+		this._url = url;
+		this._emitter = new EventEmitter();
+	}
+
+	public async connect(): Promise<void> {
+		this._ws = new WebSocket(this._url);
+
+		const connect = new Promise<void>(resolve => {
+			this._ws?.on('open', () => {
+				this.isAlive = true;
+				resolve();
+			});
+		});
+
+		const error = new Promise<void>((_, reject) => {
+			this._ws?.on('error', err => {
+				this.isAlive = false;
+				reject(err);
+			});
+		});
+
+		await Promise.race([
+			connect,
+			error,
+			timeout(CONNECTION_TIMEOUT, `Could not connect in ${CONNECTION_TIMEOUT}ms`),
+		]);
+
+		this._ws.on('ping', () => {
+			this.isAlive = true;
+		});
+
+		this._ws.on('message', data => {
+			this._handleMessage(data as string);
+		});
+	}
+
+	public async disconnect(): Promise<void> {
+		this._requestCounter = 0;
+		this._pendingRequests = {};
+
+		if (!this._ws) {
+			return Promise.resolve();
```

Comment:
if `async`, no need to specify `Proise.resolve()`. u can just say `return`?

Reply:
es-lint complaint about consistent return, so had to add this.